cobra/flux_analysis/gapfilling.py

```diff
@@ -2,8 +2,350 @@
+
+        Done incrementally so that second time the function is called,
+        active indicators in the current solutions gets higher cost than the
+        unused indicators.
+        """
+        for var in self.indicators:
+            if var not in self.costs:
+                self.costs[var] = var.cost
+            else:
+                if var._get_primal() > self.integer_threshold:
+                    self.costs[var] += var.cost
+        self.model.objective.set_linear_coefficients(self.costs)
+
+    def add_switches_and_objective(self):
+        """ Update gapfilling model with switches and the indicator objective.
+        """
+        constraints = list()
+        big_m = max(max(abs(b) for b in r.bounds)
+                    for r in self.model.reactions)
+        prob = self.model.problem
+        for rxn in self.model.reactions:
+            if not hasattr(rxn, 'gapfilling_type') or rxn.id.startswith('DM_'):
+                continue
+            indicator = prob.Variable(
+                name='indicator_{}'.format(rxn.id), lb=0, ub=1, type='binary')
+            if rxn.id in self.penalties:
+                indicator.cost = self.penalties[rxn.id]
+            else:
+                indicator.cost = self.penalties[rxn.gapfilling_type]
+            indicator.rxn_id = rxn.id
+            self.indicators.append(indicator)
+
+            # if z = 1 v_i is allowed non-zero
+            # v_i - Mz <= 0   and   v_i + Mz >= 0
+            constraint_lb = prob.Constraint(
+                rxn.flux_expression - big_m * indicator, ub=0,
+                name='constraint_lb_{}'.format(rxn.id), sloppy=True)
+            constraint_ub = prob.Constraint(
+                rxn.flux_expression + big_m * indicator, lb=0,
+                name='constraint_ub_{}'.format(rxn.id), sloppy=True)
+
+            constraints.extend([constraint_lb, constraint_ub])
+
+        self.model.add_cons_vars(self.indicators)
+        self.model.add_cons_vars(constraints, sloppy=True)
+        self.model.objective = prob.Objective(
+            Add(*self.indicators), direction='min')
+        self.update_costs()
+
+    def fill(self, iterations=1):
+        """Perform the gapfilling by iteratively solving the model, updating
+        the costs and recording the used reactions.
+
+
+        Parameters
+        ----------
+        iterations : int
+            The number of rounds of gapfilling to perform.
+
+        Returns
+        -------
+        iterable
+            A list of lists where each element is a list reactions that were
+            used to gapfill the model.
+
+        Raises
+        ------
+        RuntimeError
+            If the model fails to be validated (i.e. the original model with
+            the proposed reactions added, still cannot get the required flux
+            through the objective).
+        """
+        used_reactions = list()
+        for i in range(iterations):
```

Comment:
So I haven't really read the original papers am am not completely sure what the loop does. In case it is used to find redundant activation pathways it might benefit from a break statement in case that solution is empty (no additional reactions found). Otherwise it just reruns the same optimization several times without doing anything.